win: find and setup for VS2017

joaocgreis opened the pull request:
Reviewed https://github.com/nodejs/node-gyp/pull/1103 :
- Filter VS2017 installations using the installed components, to make sure the installation found is usable.
- Added logging.

cc @refack @bzoz can you review?

refack commented:
Personally I would move logic from the C# to JS. I like that you use more of the information available to us from the COM, but I would filter and choose in JS. I also like the idea of a log, but maybe if the C# prints everything, there will be no need.
I'll take some ideas to #1103 and to [windows-autoconf](https://github.com/node4good/windows-autoconf) 😉

joaocgreis commented:
Updated, fixed failure when there's no usable installation and store the MSBuild executable path in config so that it can respect `--msvs_version`. I considered filtering on the JS side, but it can be a huge list of components and is easy enough to do on the C# side. We can always improve if necessary.

Tested (with node v7 on Windows 7):
- Several installations of VS2017, where only one is usable
- VS2017 installed but not usable
- After uninstalling VS2017
- VSBT2015 and VS2017 side by side (`--msvs_version` works)
- Run npm as a normal user
- On Windows 10, node v4, running as a normal user

bnoordhuis commented:
Is this PR taking a dependency on powershell and the C# compiler?  That... seems unfortunate.

joaocgreis commented:
Landed in https://github.com/nodejs/node-gyp/commit/b456dd0d46a143c3e6afd63116124800f3c1c18f

@bnoordhuis it does, but those two are installed by default everywhere I've looked. To make node-gyp work out of the box in the platforms we support, it's either this, `comtypes` as in https://github.com/nodejs/node-gyp/pull/1101, or have an [executable](https://github.com/nodejs/node-gyp/pull/1103#issuecomment-282196116) committed. This worked great everywhere I've tested, but it's not impossible that there is some uncommon Windows configuration that breaks. We can fix those based on actual issues, or decide for any of the other options.

bnoordhuis commented:
Merging is a little premature.  There has been no sign-off from a committer and you didn't run the CI either.

refack commented:
LGTM(ish) I'll make another PR with some more changes 😉

joaocgreis commented:
Took it back, I took @bzoz sign-off as enough, this has been under discussion in https://github.com/nodejs/node-gyp/pull/1101 and https://github.com/nodejs/node-gyp/pull/1103 for a long time. I'll run CI.

refack commented on lib/build.js:
```diff
@@ -14,7 +14,9 @@ var fs = require('graceful-fs')
   , processRelease = require('./process-release')
-  , win = process.platform == 'win32'
+  , win = process.platform === 'win32'
+if (win)
```
You don't need this if you already injected `config.variables.msbuild_path` (which is smart)

joaocgreis replied:
Thanks!

joaocgreis commented:
CI:
- v.7.6.0: https://ci.nodejs.org/view/All/job/nodegyp-test-commit/39/
- v6.10.0: https://ci.nodejs.org/view/All/job/nodegyp-test-commit/40/
- v4.8.0: https://ci.nodejs.org/view/All/job/nodegyp-test-commit/41/

joaocgreis commented:
@bnoordhuis CI is green, do you want to review? I think this is important to have, what should I do to move this forward?

joaocgreis commented:
cc @nodejs/node-gyp @nodejs/build

joaocgreis commented:
Updated: changed the MSBuild dependency component. Installing only MSBuild does not make it usable for C++ if the workload is not installed. But installing "Visual Studio C++ core features" is enough, included MSBuild.exe and the necessary files to make it usable.

Fishrock123 commented:
@bnoordhuis, @rvagg PTAL

rvagg commented:
working on this now FYI, had to get a few other pressing things out of the way but I'm reviewing and testing this out for myself now

rvagg commented on lib/configure.js:
```diff
@@ -137,6 +139,18 @@ function configure (gyp, argv, callback) {
     // disable -T "thin" static archives by default
     variables.standalone_static_library = gyp.opts.thin ? 0 : 1
 
+    if (win && !(gyp.opts.msvs_version && gyp.opts.msvs_version !== '2017')) {
```
would you mind inverting this? reads really badly. `(win && (!gyp.opts.msvs_version || gyp.opts.msvs_version === '2017'))`

rvagg commented on lib/configure.js:
```diff
@@ -137,6 +139,18 @@ function configure (gyp, argv, callback) {
+    if (win && !(gyp.opts.msvs_version && gyp.opts.msvs_version !== '2017')) {
+      const vsSetup = findVS2017()
+      if (vsSetup) {
+        gyp.opts.msvs_version = '2015'
```
are we forcing this to make GYP happy, or is this (and the next line) a mistake?

joaocgreis replied:
We are forcing this to make GYP happy, and to avoid pulling https://chromium-review.googlesource.com/#/c/433540/ (which did not land yet) as a floating patch, since floating patches seem to be avoided here.

rvagg commented on lib/find-vs2017.js:
```diff
@@ -0,0 +1,47 @@
+
+  var vsSetup
+  try {
+    const vsSetupRaw = execSync(psQuery, { encoding: 'utf8' })
```
`var` for consistency, also given that node-gyp still works for very old versions of Node

rvagg replied:
sorry, I mean, `s/var/const` in this entire file

rvagg commented on lib/configure.js:
```diff
@@ -137,6 +139,18 @@ function configure (gyp, argv, callback) {
     variables.standalone_static_library = gyp.opts.thin ? 0 : 1
 
+    if (win && !(gyp.opts.msvs_version && gyp.opts.msvs_version !== '2017')) {
+      const vsSetup = findVS2017()
```
`s/const/var`

rvagg commented on lib/Find-VS2017.cs:
```diff
@@ -0,0 +1,262 @@
+// powershell -ExecutionPolicy Unrestricted -Version "2.0" -Command "&{Add-Type -Path Find-VS2017.cs; [VisualStudioConfiguration.Main]::Query()}"
```
does this file need source attribution and/or copyright? or is it ours?

joaocgreis replied:
This was submitted by @refack in https://github.com/nodejs/node-gyp/pull/1103 , I believe it was adapted from the python COM API by @bzoz in https://github.com/nodejs/node-gyp/pull/1101 , which in turn got the GUIDs and API from the nuget package upstream. @refack can you confim?

refack replied:
I wrote it myself. GUIDs extracted from [interop](https://www.nuget.org/packages/Microsoft.VisualStudio.Setup.Configuration.Interop/)

refack replied:
Please add
```
// Copyright 2017 - Refael Ackermann
// Distributed under MIT style license
// See accompanying file LICENSE at https://github.com/node4good/windows-autoconf
```

refack replied:
I recommend using updated file https://github.com/node4good/windows-autoconf/blob/master/tools/GetVS2017Configuration.cs

joaocgreis commented:
@rvagg updated, PTAL. About the 2015->2017 GYP hack, my comment above is hidden by GitHub:

> We are forcing this to make GYP happy, and to avoid pulling https://chromium-review.googlesource.com/#/c/433540/ (which did not land yet) as a floating patch, since floating patches seem to be avoided here.

rvagg commented:
@joaocgreis would you mind putting that comment about the gyp hack into the code so we have record of it? I can see someone coming along to "fix" it later without that context. I could even imagine myself doing that!

joaocgreis commented:
Updated with the comment.

@refack can you confirm that you wrote the `.cs` file (or let us know where you got it from)? Thanks!

bnoordhuis commented on lib/configure.js:
```diff
@@ -137,6 +139,21 @@ function configure (gyp, argv, callback) {
+        process.env['GYP_MSVS_OVERRIDE_PATH'] = vsSetup.path
+        defaults['msbuild_toolset'] = 'v141'
+        defaults['msvs_windows_target_platform_version'] = vsSetup.sdk
+        variables['msbuild_path'] = path.join(vsSetup.path, 'MSBuild', '15.0', 'Bin', 'MSBuild.exe')
```
Can you wrap this at 80 columns?

bnoordhuis commented on lib/find-vs2017.js:
```diff
@@ -0,0 +1,47 @@
+
+  var ps = 'PowerShell -ExecutionPolicy Unrestricted -Command '
+  var csFile = path.join(__dirname, 'Find-VS2017.cs')
+  var psQuery = ps + '"&{Add-Type -Path \'' + csFile + '\'; [VisualStudioConfiguration.Main]::Query()}" 2>&1'
```
Likewise.

bnoordhuis commented on lib/find-vs2017.js:
```diff
@@ -0,0 +1,47 @@
+
+  var vsSetup
+  try {
+    var vsSetupRaw = execSync(psQuery, { encoding: 'utf8' })
```
Policy decision: do we still support v0.10?  If yes, you can't use execSync(), it was added in v0.12.

That aside, is there a reason to use execSync() instead of spawnSync()?  If `__dirname` has spaces in it, execSync() will probably fail.

Lastly, can powershell always be expected to be on the path?

joaocgreis replied:
I kept the sync version from the original PR because making it async would have required deeper changes in `build.js` and `configure.js`. This is no longer true after the change to how MSBuild is found, so I'll change to async.

I made sure this works when node-gyp is installed to a location with spaces in its path.

Powershell is always in the path, unless is has been tampered with. The location of the binary seems to be stable, I'll change this to use it directly to avoid issues.

joaocgreis commented:
@bnoordhuis updated, PTAL. This now works with v0.10.

Tested, with node_gyp replaced in a node installation under `Program Files` with versions v0.10.48, v0.12.18, v4.8.0, v6.10.0 and v7.7.2. Also tested without a usable VS2017 installation and without any VS installed.

refack commented:
Please see https://github.com/nodejs/node-gyp/pull/1130/files/034a33078ce30007b74ec7b94156d4270b67e1c4#r106001050

refack commented:
I recommand using updated file https://github.com/node4good/windows-autoconf/blob/master/tools/GetVS2017Configuration.cs
Same as one that landed in [`boost/build`](https://github.com/boostorg/build/blob/develop/src/tools/vc141helper/GetVS2017Configuration.cs)
And gets proper MS style CI at [![AppVeyor](https://ci.appveyor.com/api/projects/status/u7812xmwxij7ljlh?svg=true)](https://ci.appveyor.com/project/refack/windows-autoconf), which BTW runs `node-gyp` test suite

joaocgreis commented:
Updated the COM API part of the file. I'd rather keep the code part as is here, with the filtering and the logging.

Started CI:
- v7.7.2: https://ci.nodejs.org/view/All/job/nodegyp-test-commit/42/ (green)
- v0.10.48: https://ci.nodejs.org/view/All/job/nodegyp-test-commit/43/ (red on ubuntu, confirmed the same happens on master: https://ci.nodejs.org/view/All/job/nodegyp-test-commit/44/)

@bnoordhuis should we run other versions?

bnoordhuis commented:
Maybe do v4.x and v6.x just in case?

joaocgreis commented:
- v6.10.0: https://ci.nodejs.org/view/All/job/nodegyp-test-commit/45/ (green)
- v4.8.0: https://ci.nodejs.org/view/All/job/nodegyp-test-commit/46/ (green)
- v0.12.18: https://ci.nodejs.org/view/All/job/nodegyp-test-commit/47/ (green)

rvagg commented:
@joaocgreis let's get this published today. Clean it up and get it merged and we'll publish and then PR npm to get it there too. Ping me on IRC when you're around and want to move forward.

joaocgreis commented:
Landed in https://github.com/nodejs/node-gyp/commit/ae141e19060ecb1c596a96acf9510d66b7376c18

rvagg commented:
Attempted PRs to npm (I always get seem to get this process wrong):

* lts: https://github.com/npm/npm/pull/16049
* latest: https://github.com/npm/npm/pull/16050

rvagg commented:
OK, looks like we might not be getting v4 compatibility for VS 2017 from npm unless we patch it ourselves I guess @nodejs/lts https://github.com/npm/npm/pull/16049

refack commented:
What's needed for v4 compat? Can I help?

gibfahn commented:
@refack issue is that Node 4 is going into maintenance mode soon, so there might not be another update to npm before that time. The ping to @nodejs/lts is for the LTS team to decide whether it's worth doing another v4 release to pick up the npm update.

FWIW, I think it's worth updating to support VS2017 in Node 4 (but then I won't have to do the release).

refack commented:
@gibfahn, thanks
Best I can do is make sure VS2017 changes to `node-gyp` are fully compatible with v4 (I'll add it to my CI matrix)

Fishrock123 commented:
> OK, looks like we might not be getting v4 compatibility for VS 2017 from npm unless we patch it ourselves I guess

I'm not sure that was read from the comment correctly but it *could* be the case.

gibfahn commented:
>Unless Node.js is planning on doing a new v4 release for this(?), I don't see us doing a new npm@2 release.

Based on that comment, if we can get it into Node v4, we can get it into npm. 

Node v4 goes into maintenance mode on 2017-04-01, i.e. in ten days. Whichever way we go, we need to make a decision ASAP.

refack commented:
👎 
#1151 `v3.6.0` might break a lot of `actions` out there 😟

refack commented:
@joaocgreis, @rvagg Let's fast track #1153

refack commented:
**"Bring back VS150COMNTOOLS"**
I urge everyone else to up vote at https://developercommunity.visualstudio.com/content/problem/35325/missing-vs150comntools.html

refack commented on lib/Find-VS2017.cs:
```diff
@@ -0,0 +1,268 @@
+                else if (id == "Microsoft.VisualStudio.Component.VC.Tools.x86.x64")
+                    hasVCTools = true;
+                else if (id.StartsWith(Win10SDKPrefix))
+                    Win10SDKVer = Math.Max(Win10SDKVer, UInt32.Parse(id.Substring(Win10SDKPrefix.Length)));
```
Bug found with new SDK #1144

AndrewPardoe commented:
The officially supported vswhere tool is now available as part of the install starting today with Visual Studio 15.2 preview 2: https://blogs.msdn.microsoft.com/heaths/2017/04/21/vswhere-is-now-installed-with-visual-studio-2017/. This doesn't fix all the existing installs of VS 2017 but going forward you can now detect the VS install locations and whether C++ tools are available from a scripted environment. I hope this helps your scenario.

am11 commented:
One approach is to download VSWhere as a fallback if not available and completely rely on it to find VS2017:

```powershell
# powershell
$vswherePath = "${env:ProgramFiles(x86)}\Microsoft Visual Studio\Installer\vswhere.exe";
if (!(Test-Path $vswherePath)) {
    $latestReleaseTag = $(curl https://api.github.com/repos/Microsoft/vswhere/releases/latest | `
                          ConvertFrom-Json) | % { $_.tag_name };

    $vswherePath = "$pwd\vswhere.exe";
    curl -O $vswherePath `
        https://github.com/Microsoft/vswhere/releases/download/$latestReleaseTag/vswhere.exe;
}

$vsBase = &$vswherePath -latest -products * `
              -requires Microsoft.VisualStudio.Component.VC.Tools.x86.x64 `
              -property installationPath;

if (!$vsBase) {
    throw "VS or MSVC component is not installed";
}

&"$vsBase\vc\Auxiliary\Build\vcvars64.bat";
# now link.exe, cl.exe etc. are in PATH

# or write vsvars path to stdout for the caller (MSBuild Exec task)
echo $vsBase\vc\Auxiliary\Build\vcvars64.bat;
```

However, in practice, vswhere is available in a deterministic location when we have VS or buildtools installed, then downloading is sort of moot if we don't want to support non-ProgramFiles installs (e.g. https://github.com/dotnet/corert/blob/378ad6a/src/BuildIntegration/findvcvarsall.bat).

ElectronicLove commented:
So please can you help me did you find out how to solve this problem? I've tried everything still this error.
